server/src/delta_cache.rs

```diff
@@ -0,0 +1,286 @@
+use std::collections::VecDeque;
+use unleash_types::client_features::{DeltaEvent, ClientFeature, Segment};
+
+
+
+#[derive(Debug, Clone)]
+pub struct DeltaHydrationEvent {
+    event_id: i32,
+    features: Vec<ClientFeature>,
+    segments: Vec<Segment>,
+}
+
+
+#[derive(Debug, Clone)]
+pub struct DeltaCache {
+    max_length: usize,
+    events: VecDeque<DeltaEvent>,
+    hydration_event: DeltaHydrationEvent,
+}
+
+trait GetEventId {
+    fn get_event_id(&self) -> i32;
+}
+
+impl GetEventId for DeltaEvent {
+    fn get_event_id(&self) -> i32 {
+        match self {
+            DeltaEvent::FeatureUpdated { event_id, .. }
+            | DeltaEvent::FeatureRemoved { event_id, .. }
+            | DeltaEvent::SegmentUpdated { event_id, .. }
+            | DeltaEvent::SegmentRemoved { event_id, .. }
+            | DeltaEvent::Hydration { event_id, .. } => *event_id,
+        }
+    }
+}
+
+impl DeltaCache {
+    pub fn new(hydration_event: DeltaHydrationEvent, max_length: usize) -> Self {
+        let mut cache = DeltaCache {
+            max_length,
+            events: VecDeque::new(),
+            hydration_event: hydration_event.clone(),
+        };
+        cache.add_base_event_from_hydration(hydration_event);
+        cache
+    }
+
+    fn add_base_event_from_hydration(&mut self, hydration_event: DeltaHydrationEvent) {
+        if let Some(last_feature) = hydration_event.features.last().cloned() {
+            self.add_events(vec![DeltaEvent::FeatureUpdated {
+                event_id: hydration_event.event_id,
+                feature: last_feature,
+            }]);
+        }
+    }
+
+    pub fn add_events(&mut self, events: Vec<DeltaEvent>) {
+        for event in events.into_iter() {
+            self.events.push_back(event.clone());
+            self.update_hydration_event(event);
+            if self.events.len() > self.max_length {
+                self.events.pop_front();
+            }
+        }
+    }
+
+    pub fn get_events(&self) -> &VecDeque<DeltaEvent> {
+        &self.events
+    }
+
+    pub fn is_missing_revision(&self, revision_id: u32) -> bool {
+        !self.events.iter().any(|event| event.get_event_id() == revision_id as i32)
+    }
+
+    pub fn get_hydration_event(&self) -> &DeltaHydrationEvent {
+        &self.hydration_event
+    }
+
+    fn update_hydration_event(&mut self, event: DeltaEvent) {
+        match event {
+            DeltaEvent::FeatureUpdated { event_id, feature } => {
+                if let Some(existing) = self.hydration_event.features.iter_mut().find(|f| f.name == feature.name) {
+                    *existing = feature;
+                } else {
+                    self.hydration_event.features.push(feature);
+                }
+                self.hydration_event.event_id = event_id;
```

Comment:
I think having this repeated event setting line in in 4 out of the 5 enum cases is a little fragile to future updates. You also went to all that trouble of creating a get_event_id method, so can I suggest we do this rather

1) Remove the calls to `self.hydration_event.event_id = event_id` in the match arms
2) Add a ```
        if !matches!(event, DeltaEvent::Hydration { .. }) {
            self.hydration_event.event_id = event.get_event_id();
        }
```

At the top of the method